FileInstaller/FileInstaller.cpp

```diff
@@ -0,0 +1,106 @@
+﻿#include <windows.h>
+#include <iostream>
+#include <vector>
+
+#include "FileInstaller.hpp"
+#include "FileUtils.hpp"
+
+FileInstaller::FileInstaller(std::vector<std::wstring> &file_paths, std::wstring const &installation_dir)
+	: m_file_paths(file_paths),
+	m_installation_dir(installation_dir),
+	m_is_dir_already_exists(false),
+	m_file_paths_to_clean({}) {}
+
+FileInstaller::~FileInstaller() {}
+
+void FileInstaller::install() {
+	try {
+		this->_create_installation_dir();
```

Comment:
This approach generally works, but has few drawbacks:

1. your installer class needs to implement and manage the lifetime of _all_ possible actions. For example, if we would want to add code that adds values to the Windows Registry, you would need to add more code to handle both installation and recovery. So, the more possible actions you'll have, the more cluttered this class becomes.
2. Since this class is responsible to account for each action it executed, the recovery code becomes cumbersome, e.g. you need to store the paths of files you copied, you need to remember if the folder existed before creation, etc. Since FileInstaller is reponsible for all this logic, this tends to become error prone because there's too much logic in one place, for example see the possible leak I described in a different comment.
3. reverting in a catch clause is error prone, for example you didn't account for all possible exception types (what happens if push_back throws an exception? it won't be a ```FileInstallerException ``` type and you will not revert). What if you forgot to account for all possible flows?

All these problems are solveable with RAII, if you divide this class to smaller classes each responsible for exactly the action it needs to be aware of. (note that currently you didn't use RAII at all).
Try to improve your solution

Reply:
Done.